### PR TITLE
[JENKINS-70931] Remove Prototype `Ajax.Request` usages from `hudson-behavior.js`

### PR DESCRIPTION
**Pull request by timja**
See [JENKINS-70931](https://issues.jenkins.io/browse/JENKINS-70931).

### Testing done

FormChecker has been tested with the timezone field on the user configuration page.
On page load a doCheck with no attributes is done and works, and then when selecting other timezones it also works

Help fetching has been tested by putting a breakpoint on the code and then clicking help buttons, breakpoint was hit and help areas were loaded.

Lazy map was tested by going to the build time trend and seeing the breakpoint hit on page load and the page looked correct.

Stop button by adding a breakpoint on the code, starting a freestyle build, ensuring the breakpoint was hit and that it behaved as expected.

descriptionForm by editing a freestyle project description, putting a breakpoint ensuring the code was executed behaving as expected

refreshPart was tested by adding a breakpoint and making sure it was hit when executors updated.

validateButton was tested by clicking test connection on the slack plugins global configuration, ensuring breakpoint was hit and it behaved as expected

### Proposed changelog entries

- (skip)

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7951"></a>

### CODE REVIEW

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (basil) [inline war/src/main/webapp/scripts/hudson-behavior.js:241]: In the Prototype version it was possible to call this function with `params.method` set to `get`, and this would be plumbed through to `Ajax.Request`. Now, the same request would result in a POST request. I think we should retain compatibility by plumbing this through to the Fetch API. If for some reason it is desirable to change the behavior, then the preceding `if` statement can be simplified.

**Review by basil — APPROVED** [review]
Fine with me pending the one issue I opened.

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (timja) [reply war/src/main/webapp/scripts/hudson-behavior.js:235]: Removed as it was unused since 2.285
https://github.com/search?q=org%3Ajenkinsci+checkMethod%3D%22get%22+language%3AXML&type=code&l=XML

**Review by basil — DISMISSED** [review]
Fine but there are still a lot of `checkMethod="post"` usages in plugins (including proprietary plugins) so maybe we should add a developer changelog entry and/or log a warning instructing the plugin developer to remove the dead code.

**Review by NotMyFault — APPROVED** [review]

**Review by yaroslavafenkin — COMMENTED** [review]
Looks fine to me security wise

File: test/src/test/java/hudson/tools/ZipExtractionInstallerTest.java
Comment (timja) [reply test/src/test/java/hudson/tools/ZipExtractionInstallerTest.java:99]: It was a very hacky test relying on internals of htmlunit which changed when switching to fetch I wasn’t able to replicate it and it didn’t seem like a critical test

**Comment by timja** [comment]
> log a warning instructing the plugin developer to remove the dead code

https://github.com/jenkinsci/jenkins/pull/7951/commits/8187d476344faa911640f8a5ee6863e0a7c444d3

image

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (daniel-beck) [inline war/src/main/webapp/scripts/hudson-behavior.js:710]: Is `status` defined here?

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (basil) [inline war/src/main/webapp/scripts/hudson-behavior.js:1113]: The original was printing the HTTP status text sent by the server, not the text body of the response.

```suggestion
            "<b>ERROR</b>: Failed to load help file: " + rsp.statusText;
```

**Comment by timja** [comment]
> Breaks plugins.MatrixAuthPluginTest and plugins.ScriptSecurityPluginTest in ATH.

Fixed

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (basil) [inline war/src/main/webapp/scripts/hudson-behavior.js:231]: `matrix-auth` and `azure-ad` call this with "GET". This changes the existing behavior. Not sure there is a valid justification for that.

If there is a valid justification for changing the existing behavior, then this calling convention is pretty ugly. Better to define a new two-argument method and deprecate the three-argument one, logging a warning if anyone is still calling the three-argument version and filing PRs to update consumers.

Comment (timja) [reply war/src/main/webapp/scripts/hudson-behavior.js:231]: JavaScript doesn't have function overloading.
https://stackoverflow.com/questions/456177/function-overloading-in-javascript-best-practices

We're stuck with that function with that name.
We could create a new function with a new name and use an object for parameters to make it more flexible in future

Comment (basil) [reply war/src/main/webapp/scripts/hudson-behavior.js:231]: Still not sure there is a valid justification for changing the existing behavior.

Comment (timja) [reply war/src/main/webapp/scripts/hudson-behavior.js]: core 2.285 changed the default to post, these 2 would have been missed by accident as they will be the only consumers calling it directly.

anyway I've added some compat so they will be able to set GET for now and will file separate PRs to switch them to post if that's okay?

**Review by basil — DISMISSED** [review]
This PR contains two unrelated changes and is too difficult for me to review. Requesting any changes related to the deprecation of `checkMethod` to be split into a separate PR.

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (basil) [inline war/src/main/webapp/scripts/hudson-behavior.js:724]: The old code propagated the `method` to the object passed in as the second argument of `FormChecker.sendRequest`. The new code does not.

**Comment by basil** [comment]
Still [failing ATH](https://ci.jenkins.io/job/Core/job/acceptance-test-harness/job/PR-1163/2/console) as far as I can tell.

**Comment by basil** [comment]
Code looks good. Once this gets a passing ATH run (or at least one that is no worse than trunk builds) I will be ready to approve this.

**Comment by basil** [comment]
[ATH passed](https://ci.jenkins.io/job/Core/job/acceptance-test-harness/job/PR-1163/3/console) so this is good to go from my perspective.

**Review by basil — APPROVED** [review]
Thanks for your hard work cleaning up this patch and getting ATH to pass!

---

This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (timja) [reply war/src/main/webapp/scripts/hudson-behavior.js:252]: 👉 https://github.com/jenkinsci/jenkins/pull/7986

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (balakine) [inline war/src/main/webapp/scripts/hudson-behavior.js:244]: I realize I'm late to the party, but are we cool with breaking this function? It's used in plugins and the `response` parameter in `onComplete` having no `responseText` is a bit of a surprise to all the scripts.

Comment (daniel-beck) [reply war/src/main/webapp/scripts/hudson-behavior.js:244]: Could you clarify what's being broken? What scripts?

Comment (balakine) [reply war/src/main/webapp/scripts/hudson-behavior.js:244]: Here is where I noticed it
https://github.com/jenkinsci/global-build-stats-plugin/blob/master/src/main/webapp/scripts/global-build-stats/standardFunctions.js#L39

Comment (basil) [reply war/src/main/webapp/scripts/hudson-behavior.js:244]: [JENKINS-71741](https://issues.jenkins.io/browse/JENKINS-71741) FTR

**Pull request merged**
